### ISSUE TITLE
iOS Focus problem

### ISSUE THREAD

**Issue by parigino1** [afriscic/BarcodeScanning.Native.Maui#37]
Hello!
On iOS I noticed that some devices struggle to scan a code, because they can't focus (even with TapToFocusEnabled set to True). 
This happens on devices with more than one camera, noting that they use the first one at the top, while for example an iPhone 7 (which has one camera) works perfectly.
Is there any possible solution? 

Thanks!

**Comment by afriscic** [comment]
Hello.
After quite a bit of digging I found something that could be useful: [https://stackoverflow.com/questions/49541837/swift-autofocus-expose-and-continuousautofocus-exposure-at-the-same-time](https://stackoverflow.com/questions/49541837/swift-autofocus-expose-and-continuousautofocus-exposure-at-the-same-time)
I'll try to implement something like this in next release. My main problem is that I don't have an Iphone on hand to test it so it could be a bit of a hit in the dark...

**Comment by afriscic** [comment]
Fixed in 1.3.0

**Comment by nzohrab** [comment]
I was just having this issue yesterday too and am glad to see it's fixed in this release. Awesome stuff!

**Comment by nzohrab** [comment]
Alright I've managed to fix this on my end by following: https://stackoverflow.com/questions/74381985/choosing-suitable-camera-for-barcode-scanning-when-using-avcapturedevicetypebuil

CameraViewHandler.cs
`    private void UpdateCamera()
    {
        if (_captureSession is not null)
        {
            lock (_syncLock)
            {
                _captureSession.BeginConfiguration();

                _captureSession.SessionPreset = AVCaptureSession.Preset1280x720;

                if (_captureInput is not null)
                {
                    if (_captureSession.Inputs.Length > 0 && _captureSession.Inputs.Contains(_captureInput))
                        _captureSession.RemoveInput(_captureInput);

                    _captureInput.Dispose();
                }

                _captureDevice?.Dispose();
                if (VirtualView?.CameraFacing == CameraFacing.Front)
                {
                    _captureDevice = AVCaptureDevice.GetDefaultDevice(AVCaptureDeviceType.BuiltInWideAngleCamera, AVMediaTypes.Video, AVCaptureDevicePosition.Front);
                }
                else
                {
                    _captureDevice = AVCaptureDevice.GetDefaultDevice(AVCaptureDeviceType.BuiltInWideAngleCamera, AVMediaTypes.Video, AVCaptureDevicePosition.Back);
                }
                _captureDevice ??= AVCaptureDevice.GetDefaultDevice(AVMediaTypes.Video);

                SetZoomFactor();
                if (_captureDevice is not null)
                {
                    ResetFocus();
                    _captureInput = new AVCaptureDeviceInput(_captureDevice, out _);

                    if (_captureSession.CanAddInput(_captureInput))
                        _captureSession.AddInput(_captureInput);
                }

                _captureSession.CommitConfiguration();
            }

            UpdateResolution();
        }
    }
    public void SetZoomFactor()
    {
        if (UIDevice.CurrentDevice.CheckSystemVersion(15, 0))
        {
            if (_captureDevice != null)
            {
                float deviceFieldOfView = _captureDevice.ActiveFormat.VideoFieldOfView;
                float previewFillPercentage = 0.8f;
                float minimumTargetObjectSize = 15.0f * 10.0f;
                double radians = DegreesToRadians(deviceFieldOfView);
                float filledTargetObjectSize = minimumTargetObjectSize / previewFillPercentage;
                double minimumSubjectDistance = filledTargetObjectSize / Math.Tan(radians / 2.0);

                float deviceMinimumFocusDistance = _captureDevice.MinimumFocusDistance;
                if (minimumSubjectDistance < deviceMinimumFocusDistance)
                {
                    float zoomFactor = deviceMinimumFocusDistance / (float)minimumSubjectDistance;
                    NSError error;
                    if (_captureDevice.LockForConfiguration(out error))
                    {
                        _captureDevice.VideoZoomFactor = zoomFactor;
                        _captureDevice.UnlockForConfiguration();
                    }
                }
            }
        }
    }
    private double DegreesToRadians(float degrees)
    {
        return degrees * Math.PI / 180.0;
    }
`

**Comment by nzohrab** [comment]
I've tested the above on an IPhone 7 and an IPhone 12 and it's working well in both cases. I'm unsure whether this is a suitable approach but it works for me and does seem to solve the issue.

**Comment by afriscic** [comment]
I'll reopen this issue.
I looked into it and this rabbit hole is really deep. For now I don't think zooming in is a good solution as all it does is just forces user to be further away from the barcode, and  consequently limiting FOV.

I have a couple of questions:
1. On which models is this problem the most pronounced? As I've seen on internet the newer IPhones 14 PRO and 15 PRO are most affected.
2. Does the `AVCaptureDeviceType.BuiltInTripleCamera` switch to ultrawide lens on autofocus or not? According to documentation it should automagically switch lenses according to focus distance?
3. As I don't have IPhones on hand to test it's been really hard to debug this. Can you post `_captureDevice.MinimumFocusDistance` values for different devices and different lenses?

For now i think I'll create a new BundableProperty so you can select lenses in MAUI. Only downside is that i couldn't find equivalent feature on Android...

**Comment by afriscic** [comment]
First of all thank you for sharing your solution it was really helpful to better understand the problem.
I gave a lot of thought about this problem last couple of days. It's really hard to come up with a solution that would be applicable to all platforms and all different devices.
I think I won't directly use your solution as I think that reducing FOV just to steer the user to move the phone further away because of the large focus distance is not a universal solution. But I decided to implement zoom control as it can be done in a way to be compatible with all platforms and gives an ability to select different lenses on iPhones with multiple cameras and Android phones that implement logical multi-camera setup.
More info: 
[https://stackoverflow.com/questions/69587952/iphone-13-pro-swift-avfoundation-builtintriplecamera-how-to-enable-automat](https://stackoverflow.com/questions/69587952/iphone-13-pro-swift-avfoundation-builtintriplecamera-how-to-enable-automat)
[https://developer.android.com/media/camera/camera2/multi-camera](https://developer.android.com/media/camera/camera2/multi-camera)
